sqlstats: put logging of sql stats flush signal behind vmodule

xinhaoz opened the pull request:
The sql stats flush worker may be unable to signal the sql activity update job to do work if it has not yet completed its previous run. Currently we emit a warning log if this occurs, but tracking this is replaceable with the
corresponding metric. Since the frequency of failing to signal may be high, we should put this behind vmodule.

Additionally, we should only emit the log if flush is enabled for sql stats. To accomplish this we modify the `Flush` procedure to return a boolean indicating if it did not exit before doing any work.

Fixes: #121577

Release note: None

xinhaoz committed:
sqlstats: put logging of sql stats flush signal behind vmodule

The sql stats flush worker may be unable to signal the sql
activity update job to do work if it has not yet completed
its previous run. Currently we emit a warning log if this
occurs, but tracking this is replaceable with the
corresponding metric. Since the frequency of failing to
signal may be high, we should put this behind vmodule.

Additionally, we should only emit the log if flush is enabled
for sql stats. To accomplish this we modify the `Flush` procedure
to return a boolean indicating if it did not exit before doing
any work.

Fixes: #121577

Release note: None

dhartunian commented on pkg/sql/sqlstats/persistedsqlstats/provider.go:
I generally want to avoid checking a bool in multiple parts of the codebase because then it's hard to reason about the consequences of flipping the setting.

If you think avoiding the work below is important, what do you think about `s.Flush` returning a bool telling us whether it did anything. Then we can avoid signaling when it returns `false`. Then it keeps the code chunks in sync even if the setting changes in the middle and avoids us having to think about the possibility.

Also, if we're adding conditional logic to the job we should add a test to exercise the change.

xinhaoz replied:
Good idea on the bool - done and test added. We can consider renaming `Flush` to `MaybeFlush` or something if desired.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/121578)
<!-- Reviewable:end -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121578)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @xinhaoz)

___
*[`pkg/sql/sqlstats/persistedsqlstats/provider.go` line 203 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121578#gh-1548831835:-Nv7RosfBhEBDDyQqt_w:bve4p04) ([raw file](https://github.com/cockroachdb/cockroach/blob/9123e4dd4b2c49a441c637946be3b2917224e098/pkg/sql/sqlstats/persistedsqlstats/provider.go#L203)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Good idea on the bool - done and test added. We can consider renaming `Flush` to `MaybeFlush` or something if desired.
</blockquote></details>

Rename sounds good :thumbsup: 
___
*[`pkg/sql/sqlstats/persistedsqlstats/flush_test.go` line 1272 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121578#-Nv7Rj0w-W63noo7GVaC:-Nv7Rj0w-W63noo7GVaD:b-bgll8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/3147bccb76f67a58aca3c00fa5875a9d470fc8d7/pkg/sql/sqlstats/persistedsqlstats/flush_test.go#L1272)):*
> ```Go
> 
> 	sqlStatsKnobs := sqlstats.CreateTestingKnobs()
> 	tc := serverutils.StartCluster(t, 3 /* numNodes */, base.TestClusterArgs{
> ```

do we need a 3 node cluster in test? I assume just 1 would suffice.


<!-- Sent from Reviewable.io -->

xinhaoz committed:
persistedsqlstats: rename `Flush` to `MaybeFlush`

This function is more aptly named `MaybeFlush` given that
it may not flush sql stats in certain conditions.

Epic: none

Release note: None

xinhaoz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121578)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @dhartunian)

___
*[`pkg/sql/sqlstats/persistedsqlstats/provider.go` line 203 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121578#gh-1548831835:-Nv8EaAC01JkQdpGCITU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9123e4dd4b2c49a441c637946be3b2917224e098/pkg/sql/sqlstats/persistedsqlstats/provider.go#L203)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Rename sounds good :thumbsup:
</blockquote></details>

Done.
___
*[`pkg/sql/sqlstats/persistedsqlstats/flush_test.go` line 1272 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121578#-Nv7Rj0w-W63noo7GVaC:-Nv8EcBH4i8aWHD_nLck:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3147bccb76f67a58aca3c00fa5875a9d470fc8d7/pkg/sql/sqlstats/persistedsqlstats/flush_test.go#L1272)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

do we need a 3 node cluster in test? I assume just 1 would suffice.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thank you for the edits.

> Reviewed 5 of 14 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121578#-:-Nv91oVm86FZCoNoiaaY:b-b49dvz)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @abarganier)



<!-- Sent from Reviewable.io -->

xinhaoz commented:
bors r+

xinhaoz commented:
bors r-

xinhaoz commented:
bors r+

xinhaoz commented:
blathers backport 24.1

xinhaoz commented:
blathers backport 23.2
